Homogenizing geometry API

germa89 opened the pull request:
Attemp to homogenise the `geometry` API. At the moment, different methods `keypoints`, `lines`, `areas`, etc return different objects.

With this PR, all methods should return a `pyvista.MultiBlock` object which can be easily plotted, or even iterate to. This establishes a deeper connection with the `pyvista` package.


#### Discussion in here: https://github.com/ansys/pymapdl/pull/2125#issuecomment-1621821414

germa89 commented:
@Gryfenfer97 can you have a look at the typing part? i am no expert so I would appreciate a lot your comments .

germa89 commented on src/ansys/mapdl/core/mapdl_geometry.py:
```diff
@@ -377,12 +402,12 @@ def n_keypoint(self):
 
     @property
-    def knum(self):
+    def knum(self) -> NDArray[np.integer]:
```
My IDLE (VSCode) was complaining about that, but I guess you are right.

germa89 commented on src/ansys/mapdl/core/mapdl_geometry.py:
```diff
@@ -613,7 +641,12 @@ def keypoint_select(self, items, sel_type="S", return_selected=False):
+        self,
+        items: Optional[Sequence],
+        sel_type: Optional[str] = "S",
+        return_selected: Optional[bool] = False,
```
```suggestion
        return_selected: bool = False,
```

germa89 commented on src/ansys/mapdl/core/mapdl_geometry.py:
```diff
@@ -613,7 +641,12 @@ def keypoint_select(self, items, sel_type="S", return_selected=False):
+        self,
+        items: Optional[Sequence],
+        sel_type: Optional[str] = "S",
+        return_selected: Optional[bool] = False,
```
Roger!

germa89 commented on src/ansys/mapdl/core/mapdl_geometry.py:
```diff
@@ -613,7 +641,12 @@ def keypoint_select(self, items, sel_type="S", return_selected=False):
+        items: Optional[Sequence],
+        sel_type: Optional[str] = "S",
+        return_selected: Optional[bool] = False,
+    ) -> Optional[int]:
```
Got it!

germa89 commented on src/ansys/mapdl/core/mapdl_geometry.py:
```diff
@@ -613,7 +641,12 @@ def keypoint_select(self, items, sel_type="S", return_selected=False):
+        items: Optional[Sequence],
+        sel_type: Optional[str] = "S",
+        return_selected: Optional[bool] = False,
+    ) -> Optional[int]:
```
```suggestion
    ) -> Optional[NDArray[np.int32]]:
```

germa89 commented on src/ansys/mapdl/core/mapdl_geometry.py:
```diff
@@ -690,7 +723,12 @@ def line_select(self, items, sel_type="S", return_selected=False):
+        items: Optional[Sequence],
+        sel_type: Optional[str] = "S",
+        return_selected: Optional[bool] = False,
+    ) -> Optional[int]:
```
```suggestion
    ) -> Optional[NDArray[np.int32]]:
```

germa89 commented on src/ansys/mapdl/core/mapdl_geometry.py:
```diff
@@ -767,7 +805,35 @@ def area_select(self, items, sel_type="S", return_selected=False):
+        """Get volumes from MAPDL represented as ``pyvista.MultiBlock``"""
+        return self.get_volumes(return_as_list=False)
+
+    def get_volumes(self, return_as_list: bool = True) -> Union[list, pv.MultiBlock]:
```
Mmhh.... Let's wait. It is my intention to slowly type the whole library. The `mapdl-reader` is something that should/will be deprecated.

germa89 commented on src/ansys/mapdl/core/mapdl_geometry.py:
```diff
@@ -844,7 +910,7 @@ def volume_select(self, items, sel_type="S", return_selected=False):
             return self.vnum
 
-    def _select_items(self, items, item_type, sel_type):
+    def _select_items(self, items: Sequence, item_type: str, sel_type: str) -> None:
```
Good idea indeed. I will do that!

germa89 commented:
## Discussion

This PR supposed a big change regarding the usage of the `geometry` API. The changes are the following:

| Methods              | **Old API** returns |  **New API**returns |
|----------------------|---------------------|---------------------|
| `geometry.keypoints` | np.array            | pv.MultiBlock       |
| `geometry.lines`     | pv.PolyData         | pv.MultiBlock       |
| `geometry.areas`     | pv.PolyData         | pv.MultiBlock       |
| `geometry.volumes`   | **Not existent**    | pv.MultiBlock       |

Furthermore, the `get_$ENTITY` methods have the following options:

| Old behaviour now in: |
|---|
| `geometry.get_keypoints(return_as_array=True)` <br>`geometry.get_keypoints()` returns a `pv.PolyData` object.<br>`geometry.get_keypoints(return_as_list=True)` returns a `list` object. |
| `geometry.get_lines()`<br>`geometry.get_lines(return_as_list=True)` returns a `list` object. |
| `geometry.get_areas()`<br>`geometry.get_areas(return_as_list=True)` returns a list of `pv.UnstructuredGrid` entities<br>`geometry.get_areas(merge=True)` returns a `pv.PolyData` object |
| `geometry.get_volumes()` returns a `py.PolyData` object<br>`geometry.get_volumes(return_as_list=True)` returns a list of `pv.UnstructuredGrid` entities<br>`geometry.get_volumes(merge=True)` returns a `pv.PolyData` object |
I have tried to be as consistent as possible regarding the methods names and returns.

koubaa commented:
@germa89 breaking changes like this might be a tough pill to swallow for some users (though I think we have to allow it in some cases).

maybe there can be a one-liner like:

```
mapdl.legacy_geometry = True

mapdl.geometry => returns an object with the same behavior as before.
```

germa89 commented:
> @germa89 breaking changes like this might be a tough pill to swallow for some users (though I think we have to allow it in some cases).
> 
> maybe there can be a one-liner like:
> 
> ```
> mapdl.legacy_geometry = True
> 
> mapdl.geometry => returns a legacy geometry object with the same behavior as before.
> ```
> 
> Then you can simplify the new geometry object by removing any legacy compatibility features from it.

Roger.

Fixed.

I have implemented `mapdl.legacy_geometry` property which by default is False.

Further more I have implemented `LegacyGeometry` which is a subclass of `Geometry` and overwrite the properties I changed to return the old results according the table I wrote in my previous comment.

Further even more, I have subclassed the `pv.MultiBlock` class and added a `__call__` method so if anyone try to use the old API with new one will get something like this:
```pycon
>>> mapdl.geometry.areas()
---------------------------------------------------------------------------
VersionError                              Traceback (most recent call last)
line 1
----> 1 mapdl.geometry.areas()

File ~/pymapdl/src/ansys/mapdl/core/mapdl_geometry.py:55, in PyMAPDLMultiblock.__call__(self, *args, **kwargs)
     54 def __call__(self, *args, **kwargs):
---> 55     raise VersionError(VERSION_ERROR)

VersionError: 
From 0.66.0 the new PyMAPDL geometry module does not allow calls on
`geometry.keypoints`, `geometry.lines`, or `geometry.areas`.

You can activate the old API by doing:
>>> mapdl.legacy_geometry = True

For more information visit https://mapdl.docs.pyansys.com/version/stable/user_guide/mesh_geometry.html
```

Gryfenfer97 commented on src/ansys/mapdl/core/mapdl_geometry.py:
```diff
@@ -907,3 +1479,77 @@ def _select_items(self, items, item_type, sel_type):
+        """Active lines as a pyvista.PolyData"""
+        return super().get_lines()  # type: ignore
+
+    def areas(self, quality=1, merge=False) -> pv.PolyData:
```
The return type seems to be Union[pv.PolyData, list[UnstructuredGrid]]
```suggestion
    def areas(self, quality=1, merge=False) -> Union[pv.PolyData, list[UnstructuredGrid]]:
```

Gryfenfer97 commented on src/ansys/mapdl/core/mapdl_geometry.py:
```diff
@@ -105,10 +138,10 @@ def __init__(self, mapdl):
         self._log = self._mapdl._log
 
-    def _set_log_level(self, level):
+    def _set_log_level(self, level: str) -> None:
```
the type expected here is DEBUG_LEVELS (need to import it from ansys.mapdl.core.mapdl):
```suggestion
    def _set_log_level(self, level: DEBUG_LEVELS) -> None:
```

Gryfenfer97 commented on src/ansys/mapdl/core/mapdl_geometry.py:
```diff
@@ -44,7 +77,7 @@ def merge_polydata(items):
 
 
-def get_elements_per_area(resp):
+def get_elements_per_area(resp: str) -> List[int]:
```
```suggestion
def get_elements_per_area(resp: str) -> Union[List[List[int]], List[Tuple[int, int]]:
```

It could be interesting to change the [L113](https://github.com/ansys/pymapdl/blob/feat/homogenizing-geometry-API/src/ansys/mapdl/core/mapdl_geometry.py#L113) to replace the list by a tuple to fix the return type to `List[Tuple[int, int]]` which has the advantage of having a fixed size. However this could be considered as a breaking change

germa89 commented:
Pinging @PipKat for document review

Pinging @koubaa @clatapie and @Gryfenfer97 for final review

Gryfenfer97 reviewed (approved):
Sorry to come back again with my type issues

PipKat commented on doc/source/user_guide/mesh_geometry.rst:
```diff
@@ -61,9 +90,202 @@ To view the current mesh status, you can use this code:
+     N Arrays:   3
+
+
+Or using the entity name:
```
```suggestion
Or, you can use the entity name:
```

PipKat commented on doc/source/user_guide/mesh_geometry.rst:
```diff
@@ -61,9 +90,202 @@ To view the current mesh status, you can use this code:
+.. figure:: ../images/multiblock_pic1.png
+
+
+Or you can just plot a single entity:
```
```suggestion
Or, you can plot a single entity:
```

PipKat commented on doc/source/user_guide/mesh_geometry.rst:
```diff
@@ -61,9 +90,202 @@ To view the current mesh status, you can use this code:
+Geometry
+--------
+
+Starting from PyMAPDL v0.66.0, by default, all the geometry entities are returned
+as a |pv-multiblock|  object. For instance:
```
```suggestion
In PyMAPDL 0.66.0 and later, by default, all geometry entities are returned
as a |pv-multiblock|  object. For example:
```

PipKat commented on doc/source/user_guide/mesh_geometry.rst:
```diff
@@ -61,9 +90,202 @@ To view the current mesh status, you can use this code:
+     Y Bounds    -0.008, 0.018
+     Z Bounds    -0.003, 0.015
+
+Or for example:
```
```suggestion
**Example 2**:
```

PipKat commented on doc/source/user_guide/mesh_geometry.rst:
```diff
@@ -61,9 +90,202 @@ To view the current mesh status, you can use this code:
+     Y Bounds    -0.008, 0.018
+     Z Bounds    -0.003, 0.015
+
+As you can see, in the new API it is not needed to call the entities.
```
```suggestion
As you can see, you do not need to call the entitities in the new API.
```

PipKat commented on doc/source/user_guide/mesh_geometry.rst:
```diff
@@ -61,9 +90,202 @@ To view the current mesh status, you can use this code:
+:ref:`migration_geometry_api_066`. 
+
+
+The selection now is easier, allowing you to use indexing:
```
```suggestion
The selection now is easier.

You can use indexing:
```

PipKat commented on doc/source/user_guide/mesh_geometry.rst:
```diff
@@ -61,9 +90,202 @@ To view the current mesh status, you can use this code:
+Migration to the new Geometry API
+---------------------------------
+
+Several changes need to be addressed in older to update your script
+from the old API to the new API.
```
```suggestion
Several changes must be addressed in older scripts to update them
from the old API to the new API.
```

PipKat commented on doc/source/user_guide/mesh_geometry.rst:
```diff
@@ -61,9 +90,202 @@ To view the current mesh status, you can use this code:
+Several changes need to be addressed in older to update your script
+from the old API to the new API.
+
+One of the most important is that now, you don't need to call the entities like in the old API:
```
```suggestion
One of the most important is that you no longer need to call the entities like you did in the old API:
```

PipKat commented on doc/source/user_guide/mesh_geometry.rst:
```diff
@@ -61,9 +90,202 @@ To view the current mesh status, you can use this code:
+     Z Bounds    -0.003, 0.015
+
+In addition, the type of entities returned by those methods are different.
+For comparison between the new and old geometry API, the following table is provided.
```
```suggestion
This table compares the new and old APIs:
```

germa89 replied:
I should have seen that one... 🤦

PipKat commented on doc/source/user_guide/mesh_geometry.rst:
```diff
@@ -61,9 +90,202 @@ To view the current mesh status, you can use this code:
+
++--------------------+------------------------------------------+--------------------------------------------------+
+| **Function**       | **Old API**                              | **New API**                                      |
+|                    | (Function based - Needs to called)       | (Property based - Doesn't need to be called)     |
```
```suggestion
|                    | (Function based - Must be called)        | (Property based - Doesn't need to be called)     |
```

PipKat commented on doc/source/user_guide/mesh_geometry.rst:
```diff
@@ -61,9 +90,202 @@ To view the current mesh status, you can use this code:
+
+.. _Table-2:
+
+**Table 2. Comparison between objects returned by both geometry APIs.**
```
Should "geometry" here be capitalized? Is it referring to PyGeometry? I think within the topic, you should be consistent and "officially" define the "geometry API" in the first reference in the topic but then only use the "old API" and new "API". Sometimes throwing in "geometry" and sometimes leaving it out seems awkward.

germa89 replied:
It is not referring to PyGeometry. I feel like having APIs "naked" is confusing. But yeah, I see your point.

PipKat commented on doc/source/user_guide/mesh_geometry.rst:
```diff
@@ -61,9 +90,202 @@ To view the current mesh status, you can use this code:
++--------------------+------------------------------------------+--------------------------------------------------+
+
+
+For equivalence between the new and old geometry API, the following table is provided.
```
```suggestion
This table shows the equivalence between the old and new APIs:
```

PipKat commented on examples/01-geometry/02-areas.py:
```diff
@@ -92,20 +92,23 @@
 # Area Geometry
 # ~~~~~~~~~~~~~
-# Get the VTK ``PolyData`` containing lines.  This VTK mesh can be
+# Get the VTK ``Multiblock`` containing lines.  This VTK mesh can be
 # saved or plotted.  For more details, visit `Pyvista documentation <pyvista_docs_>`_
```
```suggestion
# Get the VTK ``Multiblock`` containing lines. This VTK mesh can be
# saved or plotted.  For more information, see the `Pyvista documentation <pyvista_docs_>`_
```

PipKat commented on src/ansys/mapdl/core/mapdl_geometry.py:
```diff
@@ -120,62 +157,428 @@ def _load_iges(self):
+            return self.volumes[name]
+        else:
+            raise ValueError(
+                f"The entity named '{name}' does not exist as a geometry entity."
```
Thank you for using a complete sentence!

germa89 replied:
I didn't get that one 😅

germa89 replied:
![](https://media.giphy.com/media/h58E0JsuK3h3d8B1do/giphy.gif)

PipKat replied:
@germa89 Ha! Most message strings that developers write are sentence fragments, and the Google developer documentation style guide (and basically all others) say that message strings should be full sentences--just like yours! I was trying to praise your work!

germa89 replied:
Awww.... 

![](https://media.giphy.com/media/1zatzZDiWS73W/giphy-downsized-large.gif)

PipKat commented on src/ansys/mapdl/core/mapdl_geometry.py:
```diff
@@ -120,62 +157,428 @@ def _load_iges(self):
-        """Keypoint coordinates"""
-        return np.asarray(self._keypoints.points)
+    def keypoints(self) -> pv.MultiBlock:
+        """Obtain the keypoints geometry
```
```suggestion
        """Obtain the keypoints geometry.
```

PipKat commented on src/ansys/mapdl/core/mapdl_geometry.py:
```diff
@@ -120,62 +157,428 @@ def _load_iges(self):
+        >>> keypoint0
+        pyvista_ndarray([ 0.     ,  0.01778, -0.00318])
+
+        Or using the entity name:
```
```suggestion
        You can use the entity name:
```

PipKat commented on src/ansys/mapdl/core/mapdl_geometry.py:
```diff
@@ -120,62 +157,428 @@ def _load_iges(self):
+        >>> kp1
+        pyvista_ndarray([ 0.     ,  0.01778, -0.00318])
+
+        You can use this to iterate over the different elements:
```
What is "this"? A pronoun here can lead to confusion. Are you refering to the preceding code? If so, write "You can use the preceding code to iterate...."

PipKat commented on src/ansys/mapdl/core/mapdl_geometry.py:
```diff
@@ -120,62 +157,428 @@ def _load_iges(self):
+        return_as_array: bool = False,
+        return_ids_in_array: bool = False,
+    ) -> Union[NDArray[Any], pv.PolyData, List[pv.PolyData]]:
+        """Obtain the keypoints geometry
+
+        Obtain the selected keypoints as a :class:`pyvista.PolyData` object or
+        a list of :class:`pyvista.PolyData`
```
```suggestion
        """Obtain the keypoints geometry.

        Obtain the selected keypoints as a :class:`pyvista.PolyData` object or
        a list of :class:`pyvista.PolyData`.
```

PipKat commented on src/ansys/mapdl/core/mapdl_geometry.py:
```diff
@@ -120,62 +157,428 @@ def _load_iges(self):
+
+        Parameters
+        ----------
+        return_as_list : bool, optional
+            Return the data as a list, by default False
+        return_as_array : bool, optional
+            Return the data as a numpy array, by default False
+        return_ids_in_array : bool, optional
+            Return the keypoints ids in the array. Only valid when
+            `return_as_array` is True. By default False
```
```suggestion
        return_as_list : bool, optional
            Whether to return the data as a list, The default is ``False``.
        return_as_array : bool, optional
            Whether to return the data as a numpy array. The default  is ``False``.
        return_ids_in_array : bool, optional
            Whether to return the keypoint IDs in the array. The default is
            ``False``. This parameter is only valid when ``return_as_array=True``.
```
To help out the user, you should say what the behavior is when needed. For example:

 return_as_list : bool, optional
            Whether to return the data as a list, The default is ``False``, in which case
            the data is returned as a string.

Note that I'm guessing here because I don't know. Maybe our users won't know either?

germa89 replied:
Very valid point.

PipKat replied:
@germa89 I made some minor edits inline to your block above. I don't mind that the "in which case" is the same for two parameters in a row. At least the user knows what behavior to expect.

PipKat commented on src/ansys/mapdl/core/mapdl_geometry.py:
```diff
@@ -120,62 +157,428 @@ def _load_iges(self):
-        """Active lines as a pyvista.PolyData"""
-        return self._lines
+    def lines(self) -> pv.MultiBlock:
+        """Obtain the lines geometry
```
```suggestion
        """Lines geometry.
```
Alex told me that properties should have a verb in the description. It's basically a "noun." Here, the plural lines is confusing because it should likely be possessive "Lines' geometry," which looks awkward, or better yet "Geometry of the lines."

PipKat replied:
@germa89 Not sure what we are obtaining. I assumed that we get one geometry having several lines. If we want use "Geometry lines", then what do we get back? Is it a line's geometry (geometry of one line) or lines' geometry (geometry of several lines) or are we getting back multiple geometries of multiple lines? Do you see why I don't know what to suggest?!

germa89 replied:
I seee.... we get one geometrical object which contains the geometry of several lines...

germa89 replied:
So I guess we should go for ``Lines geometry``

germa89 replied:
Whatever, what you say it is good. :)

PipKat commented on src/ansys/mapdl/core/mapdl_geometry.py:
```diff
@@ -120,62 +157,428 @@ def _load_iges(self):
     @property
-    def _lines(self):
+    def _lines(self) -> List[pv.PolyData]:
         """Returns lines cache"""
```
```suggestion
        """Cache of the lines."""
```

germa89 replied:
I didn't know about not allowing verbs in the property descriptions. :)

PipKat commented on src/ansys/mapdl/core/mapdl_geometry.py:
```diff
@@ -120,62 +157,428 @@ def _load_iges(self):
+          Z Bounds:   -3.180e-03, -3.180e-03
+          N Arrays:   1
+
+        Or using the entity name:
```
```suggestion
        You can use the entity name:
```

PipKat commented on src/ansys/mapdl/core/mapdl_geometry.py:
```diff
@@ -120,62 +157,428 @@ def _load_iges(self):
+          Z Bounds:   -3.180e-03, -3.180e-03
+          N Arrays:   1
+
+        You can use this to iterate over the different elements:
```
Again, the "this" has no clear antecedent. You can use what? ... the preceding code?

germa89 replied:
Got it.

PipKat commented on src/ansys/mapdl/core/mapdl_geometry.py:
```diff
@@ -120,62 +157,428 @@ def _load_iges(self):
+        Parameters
+        ----------
+        return_as_list : bool, optional
+            Return the lines in a list, by default False
```
```suggestion
            Whether to return the lines in a list. The default is ``False``.
```

PipKat commented on src/ansys/mapdl/core/mapdl_geometry.py:
```diff
@@ -120,62 +157,428 @@ def _load_iges(self):
+
+    @property
+    def areas(self) -> pv.MultiBlock:
+        """Obtain the areas geometry
```
```suggestion
        """Geometry of the areas as a :class:`pyvista.MultiBlock` object.
```
Is there only one geometry of the area as written?  You can delete line 478 and put all the needed information in one description line.

germa89 replied:
In theory first line is a short description and the following text block more detailed. But I did not have much to write there.

I will keep the return object descriptor out of the short summary. It seems out of place otherwise.

PipKat commented on src/ansys/mapdl/core/mapdl_geometry.py:
```diff
@@ -120,62 +157,428 @@ def _load_iges(self):
+          Z Bounds:   -3.180e-03, -3.180e-03
+          N Arrays:   3
+
+        Or using the entity name:
```
```suggestion
        You can use the entity name:
```

PipKat commented on src/ansys/mapdl/core/mapdl_geometry.py:
```diff
@@ -120,62 +157,428 @@ def _load_iges(self):
+          Z Bounds:   -3.180e-03, -3.180e-03
+          N Arrays:   3
+
+        You can use this to iterate over the different elements:
```
The dreaded "this" pronoun again!

germa89 replied:
![](https://media.giphy.com/media/eP1fobjusSbu/giphy.gif)

PipKat commented on src/ansys/mapdl/core/mapdl_geometry.py:
```diff
@@ -120,62 +157,428 @@ def _load_iges(self):
         Parameters
         ----------
         quality : int, optional
             quality of the mesh to display.  Varies between 1 (worst)
             to 10 (best).
```
```suggestion
            Quality of the mesh to display.  Values are between 1 (worst)
            and10 (best).
```

PipKat commented on src/ansys/mapdl/core/mapdl_geometry.py:
```diff
@@ -192,28 +595,18 @@ def areas(self, quality=1, merge=False):
         """
         quality = int(quality)
         if quality > 10:
             raise ValueError("``quality`` parameter must be a value between 0 and 10")
```
No concluding period. Perhaps start this string with "The ``quality`` parameter.... so as not to start with a literal?

germa89 replied:
```suggestion
            raise ValueError("The ``quality`` parameter must be a value between 0 and 10.")
```

PipKat commented on src/ansys/mapdl/core/mapdl_geometry.py:
```diff
@@ -490,19 +889,17 @@ def _load_lines(self):
 
-    def _load_keypoints(self):
+    def _load_keypoints(self) -> Tuple[NDArray, NDArray]:
         """Load keypoints from MAPDL using IGES"""
```
```suggestion
        """Load keypoints from MAPDL using IGES."""
```

PipKat commented on src/ansys/mapdl/core/mapdl_geometry.py:
```diff
@@ -769,7 +1178,170 @@ def area_select(self, items, sel_type="S", return_selected=False):
+          Z Bounds:   -3.180e-03, 0.000e+00
+          N Arrays:   3
+
+        Or using the entity name:
```
```suggestion
        You can use the entity name:
```

PipKat commented on src/ansys/mapdl/core/mapdl_geometry.py:
```diff
@@ -769,7 +1178,170 @@ def area_select(self, items, sel_type="S", return_selected=False):
+          Z Bounds:   -3.180e-03, 0.000e+00
+          N Arrays:   3
+
+        You can use this to iterate over the different elements:
```
And "this" is what?

germa89 replied:
![](https://media.giphy.com/media/3o6MbfCTbFVOIXzoqY/giphy.gif)

PipKat commented on src/ansys/mapdl/core/mapdl_geometry.py:
```diff
@@ -769,7 +1178,170 @@ def area_select(self, items, sel_type="S", return_selected=False):
+    def get_volumes(
+        self, return_as_list: bool = False, quality: int = 4
+    ) -> Union[List[pv.PolyData], pv.PolyData]:
+        """Get active volumes from MAPDL represented as :class:`pyvista.PolyData` or
+        a list of :class:`pyvista.UnstructuredGrid`.
```
```suggestion
        """Get active volumes from MAPDL represented as a :class:`pyvista.PolyData` object
        or a list of the :class:`pyvista.UnstructuredGrid` class.
```

germa89 replied:
It is a list of ``unstructured`` objects, so I guess it is better:
```
or a list of :class:`pyvista.UnstructuredGrid` objects.
```

PipKat commented on src/ansys/mapdl/core/mapdl_geometry.py:
```diff
@@ -769,7 +1178,170 @@ def area_select(self, items, sel_type="S", return_selected=False):
+        Parameters
+        ----------
+        quality : int, optional
+            quality of the mesh to display.  Varies between 1 (worst)
+            to 10 (best).
```
```suggestion
            Quality of the mesh to display.  Values are 1 (worst)
            to 10 (best).
```

PipKat commented on src/ansys/mapdl/core/mapdl_geometry.py:
```diff
@@ -846,12 +1418,17 @@ def volume_select(self, items, sel_type="S", return_selected=False):
+        item_type: VALID_SELECTION_ENTITY_TP,
+        sel_type: VALID_SELECTION_TYPE_TP,
+    ) -> None:
         """Select items using FLST
```
```suggestion
        """Select items using FLST.
```

PipKat commented on src/ansys/mapdl/core/mapdl_geometry.py:
```diff
@@ -907,3 +1483,79 @@ def _select_items(self, items, item_type, sel_type):
+
+
+class LegacyGeometry(Geometry):
+    """Legacy pythonic representation of MAPDL CAD geometry
```
```suggestion
    """Legacy Pythonic representation of the MAPDL CAD geometry.
```

PipKat commented on src/ansys/mapdl/core/mapdl_geometry.py:
```diff
@@ -907,3 +1483,79 @@ def _select_items(self, items, item_type, sel_type):
+class LegacyGeometry(Geometry):
+    """Legacy pythonic representation of MAPDL CAD geometry
+
+    Contains advanced methods to extend geometry building and
+    selection within MAPDL.
```
```suggestion
    This class contains advanced methods for extending geometry building and
    selection within MAPDL.
```

PipKat commented on src/ansys/mapdl/core/mapdl_geometry.py:
```diff
@@ -907,3 +1483,79 @@ def _select_items(self, items, item_type, sel_type):
+    def areas(
+        self, quality=1, merge=False
+    ) -> Union[pv.PolyData, List[pv.UnstructuredGrid]]:
+        """List of areas from MAPDL represented as ``pyvista.PolyData``.
```
```suggestion
        """List of areas from MAPDL represented as a ``pyvista.PolyData`` object.
```

PipKat commented on src/ansys/mapdl/core/mapdl_geometry.py:
```diff
@@ -907,3 +1483,79 @@ def _select_items(self, items, item_type, sel_type):
+        Parameters
+        ----------
+        quality : int, optional
+            quality of the mesh to display.  Varies between 1 (worst)
+            to 10 (best).
```
```suggestion
            Quality of the mesh to display.  Values are 1 (worst)
            to 10 (best).
```

PipKat commented on src/ansys/mapdl/core/mapdl_geometry.py:
```diff
@@ -907,3 +1483,79 @@ def _select_items(self, items, item_type, sel_type):
+            to 10 (best).
+
+        merge : bool, optional
+            Option to merge areas into a single mesh. Default
+            ``False`` to return a list of areas.  When ``True``,
+            output will be a single mesh.
```
```suggestion
            Whether to merge areas into a single mesh. The default
            is ``False``, in which case a list of areas is returned.  When ``True``,
            the output is a single mesh.
```

PipKat commented on src/ansys/mapdl/core/mapdl_geometry.py:
```diff
@@ -907,3 +1483,79 @@ def _select_items(self, items, item_type, sel_type):
+            List of ``pyvista.UnstructuredGrid`` meshes representing
+            the active surface areas selected by ``ASEL``.  If
+            ``merge=True``, areas are returned as a single merged
+            UnstructuredGrid.
```
Should ``UnstructuredGrid`` be tagged as a literal, or can it be broken out to be text: unstructured grid?

germa89 replied:
Roger.

PipKat commented on src/ansys/mapdl/core/mapdl_geometry.py:
```diff
@@ -907,3 +1483,79 @@ def _select_items(self, items, item_type, sel_type):
+
+        Examples
+        --------
+        Return a list of areas as indiviudal grids
```
```suggestion
        Return a list of areas as indiviudal grids.
```

AlejandroFernandezLuces commented:
Regarding PyVista, you can use `PointSet` for points, `PolyData` for lines and `UnstructuredGrid` for any other geometry. I see that you need to separate `PolyData` and `UnstructuredGrid` because of previous issues, so no discussion there. 

My only commet is regarding the `PointSet`. If you look at the docs:
>This class is useful for improving the performance of filters on point clouds, but not plotting.

If you don't really need to use if for that reason, you could also use `PolyData` for points, to increase compatibility among returned objects. This is not a big deal anyway, so up to you 🙂

germa89 commented on doc/source/user_guide/mesh_geometry.rst:
```diff
@@ -61,9 +90,202 @@ To view the current mesh status, you can use this code:
+     N Arrays:   3
+
+
+You can now plot the different entities by just issuing plot after them:
```
No no... you got it right :). Editing...

germa89 commented on doc/source/user_guide/mesh_geometry.rst:
```diff
@@ -61,9 +90,202 @@ To view the current mesh status, you can use this code:
+     N Arrays:   3
+
+
+You can now plot the different entities by just issuing plot after them:
```
```suggestion
You can plot different entities by calling the :meth:`plot() <ansys.mapdl.core.mapdl_geometry.Geometry.keypoints.plot>` method:
```

germa89 commented on doc/source/user_guide/mesh_geometry.rst:
```diff
@@ -61,9 +90,202 @@ To view the current mesh status, you can use this code:
+
+.. _Table-2:
+
+**Table 2. Comparison between objects returned by both geometry APIs.**
```
```suggestion
**Table 2. Comparison between objects returned by both APIs.**
```

germa89 commented on examples/01-geometry/00-keypoints.py:
```diff
@@ -67,11 +67,20 @@
+###############################################################################
+# Keypoints geometry
+# ~~~~~~~~~~~~~~~~~~
+# Get the VTK ``MultiBlock`` containing keypoints.  This VTK mesh can be
+# saved or plotted.  For more details, visit `Pyvista documentation <pyvista_docs_>`_
```
I hope I will remember in the future!

germa89 commented on src/ansys/mapdl/core/mapdl_geometry.py:
```diff
@@ -29,8 +40,34 @@
+You can activate the old API by doing:
+>>> mapdl.legacy_geometry = True
+
+For more information visit https://mapdl.docs.pyansys.com/version/stable/user_guide/mesh_geometry.html
```
Truee... but this is an error message, I don't think Sphinx works here.

germa89 replied:
Double checking with doc specialist @Revathyvenugopal162

germa89 commented on src/ansys/mapdl/core/mapdl_geometry.py:
```diff
@@ -120,62 +157,428 @@ def _load_iges(self):
+          Z Bounds:   -3.180e-03, -3.180e-03
+          N Arrays:   1
+
+        You can use this to iterate over the different elements:
```
```suggestion
        You can iterate the different elements:
```

germa89 commented on src/ansys/mapdl/core/mapdl_geometry.py:
```diff
@@ -120,62 +157,428 @@ def _load_iges(self):
+          Z Bounds:   -3.180e-03, -3.180e-03
+          N Arrays:   3
+
+        You can use this to iterate over the different elements:
```
```suggestion
        You can iterate the different elements:
```

germa89 commented on src/ansys/mapdl/core/mapdl_geometry.py:
```diff
@@ -769,7 +1178,170 @@ def area_select(self, items, sel_type="S", return_selected=False):
+          Z Bounds:   -3.180e-03, 0.000e+00
+          N Arrays:   3
+
+        You can use this to iterate over the different elements:
```
```suggestion
        You can iterate the different elements:
```

germa89 commented on src/ansys/mapdl/core/mapdl_geometry.py:
```diff
@@ -907,3 +1483,79 @@ def _select_items(self, items, item_type, sel_type):
+            List of ``pyvista.UnstructuredGrid`` meshes representing
+            the active surface areas selected by ``ASEL``.  If
+            ``merge=True``, areas are returned as a single merged
+            UnstructuredGrid.
```
```suggestion
            ``UnstructuredGrid``.
```

germa89 replied:
```suggestion
            :class:`pyvista.UnstructuredGrid <pyvista.UnstructuredGrid>`.
```

germa89 commented on src/ansys/mapdl/core/mapdl_geometry.py:
```diff
@@ -907,3 +1483,79 @@ def _select_items(self, items, item_type, sel_type):
+        Returns
+        -------
+        list of pyvista.UnstructuredGrid
+            List of ``pyvista.UnstructuredGrid`` meshes representing
```
```suggestion
            List of :class:`pyvista.UnstructuredGrid <pyvista.UnstructuredGrid>` meshes representing
```

germa89 commented on doc/source/user_guide/mesh_geometry.rst:
```diff
@@ -61,9 +90,208 @@ To view the current mesh status, you can use this code:
+     Y Bounds    -0.008, 0.018
+     Z Bounds    -0.003, 0.015
+
+As you can see, you do not need to call the entitities in the new API.
```
```suggestion
As you can see, you do not need to call the entities in the new API.
```

germa89 commented:
> Regarding PyVista, you can use `PointSet` for points, `PolyData` for lines and `UnstructuredGrid` for any other geometry. I see that you need to separate `PolyData` and `UnstructuredGrid` because of previous issues, so no discussion there.
> 
> My only commet is regarding the `PointSet`. If you look at the docs:
> 
> > This class is useful for improving the performance of filters on point clouds, but not plotting.
> 
> If you don't really need to use if for that reason, you could also use `PolyData` for points, to increase compatibility among returned objects. This is not a big deal anyway, so up to you 🙂

I dont think we do much filtering with keypoints but we do a lot of plotting. So I guess choosing `pv.PolyData` is good and consistent with the homogenisation perspective.

Thank you for your feedback!

germa89 commented:
@clatapie
